### PR TITLE
Analyser integration

### PR DESCRIPTION
**Pull request by xavier-martin**
Implement a clustering and anomaly-detection CLI tool.

### CODE REVIEW

File: bin/analysercli [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+from ivre.cliutils import COLORS, StackCli
+from ivre.analyser import Context
+
+
+# --------------------------------------------------------------------------- #
+
+
+class AnalyserCli(StackCli):
+    """
+    Base class for stacked contexts CLI
```
Comment (p-l-) [inline bin/analysercli]: How is it different from its parent?

File: bin/analysercli [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+
+class AnalyserCli(StackCli):
+    """
+    Command line utility wrapper to access analyser API
+    """
+    # COLORIZE levels
+    LOG_USAGE = 0
+    LOG_INFO = 10
+    LOG_WARNING = 20
```
Comment (p-l-) [inline bin/analysercli]: Maybe you could use a `dict` like:
```
LOG_LEVELS = {
    LOG_USAGE: {"name": "USAGE", "color": "green"},
    [...]
}
```

That could make the code in `_colorized_log()` better.

Also, what about moving that to the `cliutils` sub-module?

Comment (xavier-martin) [inline bin/analysercli]: That's a good idea !

File: ivre/analyser.py [hunk 3]
```diff
@@ -0,0 +1,667 @@
...
+        """
+        min_size = []
+        max_size = []
+        dim = hosts.shape[1]
+        for idx in xrange(dim):
+            min_size.append(round(hosts[:, idx].min() - 1))
+            max_size.append(round(hosts[:, idx].max() + 1))
+        return (min_size, max_size)
+
+    def plot(self, dim=2):
```
Comment (p-l-) [inline ivre/analyser.py]: If the only valid values for `dim` are 2 and 3, use a boolean like `plot3d` with default value `False`.

File: bin/analysercli [hunk 4]
```diff
@@ -0,0 +1,263 @@
+#! /usr/bin/env python
+# -*- coding:utf-8 -*-
```
Comment (p-l-) [inline bin/analysercli]: Why is this needed?

Comment (xavier-martin) [inline bin/analysercli]: Is it not a good practice to always start the code with this statement ?

Comment (p-l-) [inline bin/analysercli]: If all your code is plain ASCII, I don't understand why you would need that.

File: bin/analysercli [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+        context.dfilter = self._dfilter
+        context.init_hosts()
+        self.push(context)
+
+    def postloop(self):
+        colorize_log('info', 'Leaving..')
+
+    @staticmethod
+    def flt_param(param):
+        """Search for parameter starting with '!'"""
```
Comment (p-l-) [inline bin/analysercli]: 1. The function name and the docstring are not really consistent with what the function actually does.
2. If this function is only used from a "private" method, it should probably also be "private", or have I missed something?
3. Please avoid unneeded variables:
```Python
if param.startswith('!'):
    return True, param[1:]
return False, param
```

Or better:
```Python
return (True, param[1:]) if param.startswith('!') else (False, param)
```

Comment (xavier-martin) [inline bin/analysercli]: You're right, it's shorter thus better, thanks !

File: bin/analysercli [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        """Help for do_count command"""
+        log = 'count\n'
+        log += '\tPrint host count to sdout'
+        colorize_log('usage', log)
+
+    @staticmethod
+    def help_plot2d():
+        """Print help for do_plot2d command"""
+        log = 'plot2d\n'
+        log += '\tPlot hosts in two dimensions'
```
Comment (p-l-) [inline bin/analysercli]: This is bad. You need to avoid `+` on strings (particularly when they're useless). Here write:
```Python
log = """plot2d                                                                 
\tPlot hosts in two dimensions"""
```
or:
```Python
log = "plot2d\n\tPlot hosts in two dimensions"
```
(both are fine for me)

File: bin/analysercli [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+# (at your option) any later version.
+#
+# IVRE is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
+# License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with IVRE. If not, see <http://www.gnu.org/licenses/>.
+
```
Comment (p-l-) [inline bin/analysercli]: Please add a module docstring.

File: bin/analysercli [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+                self._dfilter['limit'] = int(value)
+            else:
+                query.append((neg, param, value))
+        (self._dfilter['filter'], _, _, _,
+         self._dfilter['skip'], _) = webutils.flt_from_query(
+             query, base_flt=db.nmap.flt_empty)
+
+    def do_count(self, _):
+        """Print host count to stdout"""
+        log = 'Number of hosts:\n'
```
Comment (p-l-) [inline bin/analysercli]: Here again to avoid `+` on strings:
```Python
colorize_log('info', "Number of hosts:\n%s" % ''.join(
    '\t%s:\t%s\n' % (key, value)
    for key, value in self.get_context().count().iteritems()
))
```
or (better?):
```Python
colorize_log('info', "Number of hosts:\n\t%s\n" % '\n\t'.join(
    '%s:\t%s' % (key, value)
    for key, value in self.get_context().count().iteritems()
))
```

File: bin/analysercli (hunk 6 above)
Comment (xavier-martin) [inline bin/analysercli]: Ok, noted !

File: bin/analysercli (hunk 7 above)
Comment (xavier-martin) [inline bin/analysercli]: Done.

File: ivre/analyser.py [hunk 9]
```diff
@@ -0,0 +1,675 @@
...
+        """
+        Return a new `Context` which only contains cluster centers
+        Depending on self._clusters type, centers might be one host
+        per cluster or every core-cluster hosts
+        """
+        if self._clusters is None:
+            raise RuntimeError('self._clusters must be initialised first')
+        new_list_id, new_matrix = self._clusters.select_centers(self._list_id,
+                                                                self._matrix)
+        cmd_name = 'sel-centers-' + str(self._clusters)
```
Comment (p-l-) [inline ivre/analyser.py]: Please use `'sel-centers-%s' % self._clusters` instead.

File: ivre/analyser.py [hunk 10]
```diff
@@ -0,0 +1,675 @@
...
+        """
+        Plot n_comp most relevant components for the first dimensions
+        """
+        assert dimension <= 3
+        # Prevent from index error in case PCA went on fewer dimension
+        n_comp = min(n_comp, len(pca_indexes[0]), len(pca_indexes[1]))
+        fig = plt.figure()
+        AXIS = ['X', 'Y', 'Z']
+        for dim in xrange(dimension):
+            plot = fig.add_subplot(dimension, 1, dim+1)
```
Comment (p-l-) [inline ivre/analyser.py]: PEP8: `dim + 1`

**Comment by p-l-** [comment]
Thanks for your work @xavier-martin!
